### ISSUE TITLE
How do i get FPS 

### ISSUE THREAD

**Issue by ONNONS** [ultralytics/yolov5#4176]
## ❔Question

I tried modifying the detect.py code to find out the fps, but gave up. However, since I know the video frame and the question time, I am wondering if I can know the fps value with these numbers, and I would appreciate it if you tell me where and how to add the fps to the detect.py code. This is to check the fps in real time.

## Additional context
![image](https://user-images.githubusercontent.com/65228677/127112343-2bfcc8c3-e3c7-4159-b213-0c8e25512fa4.png)

**Comment by SamSamhuns** [comment]
In `detect.py`, just add the lines denoted in `+` to get the average FPS for the complete operation set (inference + nms + post_process + save)

```python
     if pt and device.type != 'cpu':
         model(torch.zeros(1, 3, imgsz, imgsz).to(device).type_as(next(model.parameters())))  # run once
     t0 = time.time()
+    frame_count = 0
     for path, img, im0s, vid_cap in dataset:
+        frame_count += 1
         if pt:
             img = torch.from_numpy(img).to(device)
             img = img.half() if half else img.float()  # uint8 to fp16/32
@@ -187,6 +189,9 @@ def run(weights='yolov5s.pt',  # model.pt path(s)
                         vid_writer[i] = cv2.VideoWriter(save_path, cv2.VideoWriter_fourcc(*'mp4v'), fps, (w, h))
                     vid_writer[i].write(im0)

+    avg_fps = frame_count / (time.time()-t0)
+    print("Printing Average FPS for inference + nms + postprocess + saving_results")
+    print(f"Average FPS: {avg_fps:.2f}")
     if save_txt or save_img:
         s = f"\n{len(list(save_dir.glob('labels/*.txt')))} labels saved to {save_dir / 'labels'}" if save_txt else ''
         print(f"Results saved to {save_dir}{s}")
```

**Comment by ONNONS** [comment]
> 에서 `detect.py`, 그냥 표시된 줄을 추가 `+`전체 조작 세트의 평균 FPS를 얻기 위해 (추론 + NMS + post_process + 저장)
> 
> ```python
>      만약  PT  및  장치 . 유형  =!  'CPU' :
>           모델 ( 토치 . 제로 ( 1 , 3 , imgsz , imgsz ). 에 ( 장치 ). type_as ( 다음 ( 모델 . 매개 변수 ())))   번 # 실행 
>      T0  =  시간 . 시간 ()
>  +     frame_count  =  0 
>      에 대한  경로 , IMG , im0s, vid_cap  의  세트 :
>  +         frame_count  + =  1 
>          의 경우  PT :
>               IMG  =  토치 . from_numpy ( img ). 행 ( 장치 )
>               IMG  =  IMG . half () if  half  else  img . float ()   # uint8에서 fp16/32 
> @@ - 187 , 6  + 189 , 9 @@ DEF  실행 ( 가중치 = 'yolov5s.pt' ,   # model.pt의 경로 (S) 
>                          vid_writer [ I ] =  CV2 . VideoWriter ( save_path , CV2 . VideoWriter_fourcc ( * 'MP4V' ), FPS ( W , H ))
>                       vid_writer [ i ]. 쓰기 ( im0 )
> 
> +     avg_fps  =  frame_count  / ( time . time () - t0 )
>  +     print ( "추론을 위한 평균 FPS 인쇄 + nms + 후처리 + save_results" )
>  +     print ( f"평균 FPS: { avg_fps :.2f } " )
>       if  save_txt  또는  save_img :
>           s  =  f" \n { len ( list ( save_dir . glob ( 'labels/*.txt'))) } 라벨은 { save_dir  /  'labels' } "  if  save_txt  else  '' 
>          print ( f"결과는 { save_dir } { s } " 에 저장됨 )
> ```

Thanks so much, I solved it, just one more question Can you tell me how to put fps put text on the top of the output video?

**Comment by SamSamhuns** [comment]
Just add the section between the +++++++ .... ++++++++

```python
# Print time (inference + NMS)
print(f'{s}Done. ({t2 - t1:.3f}s)')

#++++++++
# add inf_nms fps info on each frame
inf_nms_post_proc_fps = time.time() - t1
cv2.putText(im0, f'FPS: {1/inf_nms_post_proc_fps:.2f}', (2, int(im0.shape[0]/35)),
            cv2.FONT_HERSHEY_SIMPLEX, fontScale=1,
            color=(0, 0, 255), thickness=2, lineType=cv2.LINE_AA)
#++++++++

# Stream results
if view_img:
    cv2.imshow(str(p), im0)
    cv2.waitKey(1)  # 1 millisecond
```

**Comment by ONNONS** [comment]
> ++++++++ .... ++++++++ 사이에 섹션을 추가하기만 하면 됩니다.
> 
> ```python
> # 인쇄 시간 (추론 + NMS) 
> print ( f' { s } 완료. ( { t2  -  t1 :.3f } s)' )
> 
> #++++++++ 
> # 각 프레임에 inf_nms fps 정보 추가 
> inf_nms_post_proc_fps  =  time . 시간 () -  t1 
> cv2 . putText ( IM0 , f'FPS : { 1 / inf_nms_post_proc_fps : .2f } ' ( 2 , INT ( IM0 . 모양 [ 0 ] / 35 )),
>              CV2 . FONT_HERSHEY_SIMPLEX , fontScale = 1 ,
>              컬러 =( 0 , 0 , 255 ), 두께 = 2 , 선 종류 = CV2 . LINE_AA )
>  #++++++++
> 
> # 스트림 결과 
> 의 경우  view_img :
>      CV2 . imshow ( STR ( P ) IM0 )
>      CV2 . waitKey ( 1 )   # 1밀리초
> ```

thank you so much
May I ask one more question?
I need to explain the output data in yolo v5

1.confusion_matrix.png
2.F1_curve.png
3.labels_correlogram.jpg
4.labels.jpg
5.P_curve.png
6.PR_curve.png
7.R_curve.png
8.Result.png

I can't even google it, and I'm 17, so I guess it's because I don't understand yet.

I need help

**Comment by SamSamhuns** [comment]
For those metrics, you should refer to the `val.py` file in the Yolo repo

**Comment by seungyong0313** [comment]
> In `detect.py`, just add the lines denoted in `+` to get the average FPS for the complete operation set (inference + nms + post_process + save)
> 
> ```python
>      if pt and device.type != 'cpu':
>          model(torch.zeros(1, 3, imgsz, imgsz).to(device).type_as(next(model.parameters())))  # run once
>      t0 = time.time()
> +    frame_count = 0
>      for path, img, im0s, vid_cap in dataset:
> +        frame_count += 1
>          if pt:
>              img = torch.from_numpy(img).to(device)
>              img = img.half() if half else img.float()  # uint8 to fp16/32
> @@ -187,6 +189,9 @@ def run(weights='yolov5s.pt',  # model.pt path(s)
>                          vid_writer[i] = cv2.VideoWriter(save_path, cv2.VideoWriter_fourcc(*'mp4v'), fps, (w, h))
>                      vid_writer[i].write(im0)
> 
> +    avg_fps = frame_count / (time.time()-t0)
> +    print("Printing Average FPS for inference + nms + postprocess + saving_results")
> +    print(f"Average FPS: {avg_fps:.2f}")
>      if save_txt or save_img:
>          s = f"\n{len(list(save_dir.glob('labels/*.txt')))} labels saved to {save_dir / 'labels'}" if save_txt else ''
>          print(f"Results saved to {save_dir}{s}")
> ```

I've been very, very, very helpful, thanks to the method you gave me. But now in yolov5- 6.0 version, the way you do it doesn't work. Can you tell me again? I'm sorry so so... I'm sorry, but please help me just once...
